rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,311 @@
+            }),
+        )
+        .await
+    }
+
+    async fn upsert_expired_lock(
+        &self,
+        existing: &LockItem,
+        new: &LockItem,
+    ) -> Result<LockItem, DynamoError> {
+        self.put_item(
+            new,
+            Some(expressions::PK_EXISTS_AND_RVN_IS_THE_SAME.to_string()),
+            Some(hashmap! {
+              vars::PK_PATH.to_string() => PARTITION_KEY_NAME.to_string(),
+              vars::RVN_PATH.to_string() => RECORD_VERSION_NUMBER.to_string(),
+            }),
+            Some(hashmap! {
+                vars::RVN_VALUE.to_string() => attr(&existing.record_version_number)
+            }),
+        )
+        .await
+    }
+
+    async fn put_item(
+        &self,
+        item: &LockItem,
+        condition_expression: Option<String>,
+        expression_attribute_names: Option<HashMap<String, String>>,
+        expression_attribute_values: Option<HashMap<String, AttributeValue>>,
+    ) -> Result<LockItem, DynamoError> {
+        let last_updated = now_millis();
+
+        self.client
+            .put_item(PutItemInput {
+                table_name: self.table_name.to_string(),
+                item: hashmap! {
+                    PARTITION_KEY_NAME.to_string() => attr(PARTITION_KEY_VALUE),
+                    OWNER_NAME.to_string() => attr(&item.owner_name),
+                    RECORD_VERSION_NUMBER.to_string() => attr(&item.record_version_number),
+                    LEASE_DURATION.to_string() => attr(&item.lease_duration),
+                },
+                condition_expression,
+                expression_attribute_names,
+                expression_attribute_values,
+                ..Default::default()
+            })
+            .await?;
+
+        Ok(LockItem {
+            is_released: false,
+            lookup_time: last_updated,
+            ..item.clone()
+        })
+    }
+}
+
+impl LockItem {
+    fn new() -> Self {
+        Self {
+            owner_name: String::from("TEST"),
+            record_version_number: Uuid::new_v4().to_string(),
+            lease_duration: 20,
+            is_released: false,
+            lookup_time: now_millis(),
+        }
+    }
+
+    fn is_expired(&self) -> bool {
+        if self.is_released {
+            return true;
+        }
+        println!(
+            "is_expired = {} - {} ({}) > {}",
+            now_millis(),
+            self.lookup_time,
+            now_millis() - self.lookup_time,
+            (self.lease_duration as u128) * 1000
+        );
+        now_millis() - self.lookup_time > (self.lease_duration as u128) * 1000
+    }
+}
+
+fn now_millis() -> u128 {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis()
+}
+
+fn attr<T: ToString>(s: T) -> AttributeValue {
+    AttributeValue {
+        s: Some(s.to_string()),
+        ..Default::default()
+    }
+}
+
+impl From<RusotoError<GetItemError>> for DynamoError {
```

Comment:
You should be able to use the `#[from]` attribute from `thiserror` to skip this implementation